src/encoding/addition.py

```diff
@@ -0,0 +1,23 @@
+from typing import Sequence
+
+from .utils import is_circuit_code
+from .typing import U8
+
+
+@is_circuit_code
+def check_add(
+    a8s: Sequence[U8],
+    b8s: Sequence[U8],
+    sum8s: Sequence[U8],
+    carry: Sequence[bool],
+):
+    assert len(a8s) == len(b8s) == len(sum8s) == 32
+    assert len(carry) == 32
+
+    # Before we add anything yet, the carry bit should be 0
+    # We allow the last carry bit to be 1 to support the overflow case
+    carry = [0] + carry[:]
+
+    for i in range(32):
+        assert carry[i + 1] * (carry[i + 1] - 1) == 0, "carry should be 0 or 1"
```

Comment:
Should we check from `0..31`?
```suggestion
        assert carry[i] * (carry[i] - 1) == 0, "carry should be 0 or 1"
```

Reply:
Since we add a [0] to carry at line 19 so that the carry have length 33 and thus we check 1..32. 

But that's confusing indeed, I shouldn't mutate the input argument carry. Let me fix it.

Reply:
Sorry I just noticed on line 76 which makes carry to length 32 and totally missed line 19... It makes sense now